packages/contracts-bedrock/src/legacy/FeeVaultInitializer.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { FeeVault } from "src/L2/FeeVault.sol";
+import { BaseFeeVault } from "src/L2/BaseFeeVault.sol";
+import { SequencerFeeVault } from "src/L2/SequencerFeeVault.sol";
+import { L1FeeVault } from "src/L2/L1FeeVault.sol";
+import { OperatorFeeVault } from "src/L2/OperatorFeeVault.sol";
+
+// Libraries
+import { Types } from "src/libraries/Types.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+
+// Interfaces
+import { ISemver } from "interfaces/universal/ISemver.sol";
+import { IProxy } from "interfaces/universal/IProxy.sol";
+
+/// @title FeeVaultInitializer
+/// @notice This contract migrates the fee vaults from the legacy system to the new system.
+contract FeeVaultInitializer is ISemver {
+    /// @notice Semantic version.
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    /// @notice Constructs the FeeVaultMigrator contract.
+
+    function migrate() public {
+        _migrateBaseFeeVault();
+        _migrateSequencerFeeVault();
+        _migrateL1FeeVault();
+        _migrateOperatorFeeVault();
+    }
+
+    function _migrateBaseFeeVault() internal {
+        // Grab current values from the old fee vault
+        address currentBaseFeeVaultRecipient = FeeVault(payable(Predeploys.BASE_FEE_VAULT)).recipient();
+        Types.WithdrawalNetwork currentBaseFeeVaultWithdrawalNetwork =
+            FeeVault(payable(Predeploys.BASE_FEE_VAULT)).withdrawalNetwork();
+        uint256 currentBaseFeeVaultMinWithdrawalAmount =
+            FeeVault(payable(Predeploys.BASE_FEE_VAULT)).minWithdrawalAmount();
+
+        // Withdraw funds from the old fee vaults
+        FeeVault(payable(Predeploys.BASE_FEE_VAULT)).withdraw();
+
+        // Deploy new implementation
+        BaseFeeVault newBaseFeeVault = new BaseFeeVault();
+
+        // Upgrade the proxy and initialize the new implementation
+        IProxy(payable(Predeploys.BASE_FEE_VAULT)).upgradeToAndCall(
```

Comment:
I am not sure if this is going to work exactly because it needs to be the proxy admin or `address(0)` that makes this call, see https://github.com/ethereum-optimism/optimism/blob/7c9586773f980266cce7235f38cc6ee7295a8870/packages/contracts-bedrock/src/universal/Proxy.sol#L75

What we usually do is have a network upgrade transaction that spoofs `address(0)` to make the initialize call. You can see an example here: https://github.com/ethereum-optimism/optimism/blob/7c9586773f980266cce7235f38cc6ee7295a8870/op-node/rollup/derive/isthmus_upgrade_transactions.go#L105

Reply:
I was thinking there has to be some way that its possible to thread the information to the right place. Looking at the code now, the easiest thing to do would be to just keep the values as immutables because we could read the values, deploy the fee vaults with immutables and then follow up with a transaction that spoofs `address(0)` and calls the proxy `upgradeToAndCall`

Moving away from immutables is just an optimization for the runbook, if its impossible to migrate to them being in storage that is ok